File: packages/backend/src/main/kotlin/quizzem/backend/features/question/service/QuestionService.kt
package quizzem.backend.features.question.service

import jakarta.persistence.EntityNotFoundException
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import quizzem.backend.core.api.dto.PageableDto
import quizzem.backend.features.question.api.dto.CreateQuestionDto
import quizzem.backend.features.question.api.dto.GetAllQuestionsParamsDto
import quizzem.backend.features.question.api.dto.UpdateQuestionDto
import quizzem.backend.features.question.model.AnswerMode
import quizzem.backend.features.question.model.AnswerModel
import quizzem.backend.features.question.model.MediaType
import quizzem.backend.features.question.model.QuestionModel
import quizzem.backend.features.question.repository.QuestionRepository
import java.util.*

@Service
class QuestionService(
    val questionRepository: QuestionRepository,
) {
    fun getAllQuestions(params: GetAllQuestionsParamsDto): PageableDto<QuestionModel> {
        val pageable = PageRequest.of(params.page ?: 0, params.size ?: 10)
        val page = questionRepository.findAll(pageable)

        return PageableDto(
            data = page.content,
            page = pageable.pageNumber,
            size = pageable.pageSize,
            totalElements = page.totalElements,
            totalPages = page.totalPages
        )
    }

    fun getQuestionById(id: UUID): QuestionModel {
        return questionRepository.findById(id).orElseThrow({ EntityNotFoundException("No question with id $id") })
    }

    fun createQuestion(dto: CreateQuestionDto) {
        val newQuestion = QuestionModel(
            text = dto.text,
            answerMode = dto.answerMode ?: AnswerMode.SINGLE_CHOICE,
            mediaType = dto.mediaType ?: MediaType.NONE,
        )

        newQuestion.answers = dto.answers.map { answer ->
            AnswerModel(
                text = answer.text,
                isCorrectAnswer = answer.isCorrectAnswer ?: false,
                question = newQuestion
            )
        }.toMutableList()

        questionRepository.save(newQuestion)
    }

    @Transactional
    fun createQuestionsBulk(dtos: List<CreateQuestionDto>) {
        dtos.forEach { createQuestion(it) }
    }

    @Transactional
    fun updateQuestion(id: UUID, dto: UpdateQuestionDto) {
        val question =
            questionRepository.findById(id).orElseThrow { EntityNotFoundException("Question with id $id not found") }

        dto.text?.let { question.text = it }
        dto.answerMode?.let { question.answerMode = it }
        dto.mediaType?.let { question.mediaType = it }
        dto.answers?.let { incoming ->
            question.answers.clear()
            question.answers.addAll(incoming.map { answer ->
                AnswerModel(
                    text = answer.text,
                    isCorrectAnswer = answer.isCorrectAnswer,
                    question = question
                )
            })
        }
    }

    fun deleteQuestion(id: UUID) {
        questionRepository.deleteById(id)
    }


}

File: packages/backend/src/features/question/service/question.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { UUID } from 'node:crypto';
import { PageableDto } from 'src/core/api/pageable.dto';
import { CreateQuestionDto } from 'src/features/question/api/dto/create-question.dto';
import { GetAllQuestionsParamsDto } from 'src/features/question/api/dto/get-all-questions-params.dto';
import { UpdateQuestionDto } from 'src/features/question/api/dto/update-question.dto';
import { AnswerModeEnum } from 'src/features/question/model/answer-mode';
import { AnswerModel } from 'src/features/question/model/answer.model';
import { MediaTypeEnum } from 'src/features/question/model/media-type';
import { QuestionModel } from 'src/features/question/model/question.model';
import { DataSource, Repository } from 'typeorm';

@Injectable()
export class QuestionService {
  constructor(
    @InjectRepository(QuestionModel)
    private readonly questionRepository: Repository<QuestionModel>,
    private readonly dataSource: DataSource,
  ) {}

  async getAllQuestions(
    params: GetAllQuestionsParamsDto,
  ): Promise<PageableDto<QuestionModel>> {
    const [records, count] = await this.questionRepository.findAndCount({
      skip: params.page * params.size,
      take: params.size,
      relations: {
        answers: true,
      },
    });

    return new PageableDto({
      data: records,
      page: params.page,
      size: params.size,
      totalElements: count,
    });
  }

  async getQuestionById(id: UUID): Promise<QuestionModel> {
    const record = await this.questionRepository.findOne({
      where: { id },
      relations: { answers: true },
    });
    if (!record) {
      throw new NotFoundException(`Question with id ${id} not found`);
    }
    return record;
  }

  async createQuestion(dto: CreateQuestionDto): Promise<void> {
    const newQuestion = new QuestionModel();
    newQuestion.text = dto.text;
    newQuestion.answerMode = dto.answerMode ?? AnswerModeEnum.SINGLE_CHOICE;
    newQuestion.mediaType = dto.mediaType ?? MediaTypeEnum.NONE;

    newQuestion.answers = dto.answers.map((answer) => {
      const newAnswer = new AnswerModel();
      newAnswer.text = answer.text;
      newAnswer.isCorrectAnswer = answer.isCorrectAnswer ?? false;
      newAnswer.question = newQuestion;
      return newAnswer;
    });

    await this.questionRepository.save(newQuestion);
  }

  async createQuestionsBulk(dtos: CreateQuestionDto[]): Promise<void> {
    await this.dataSource.transaction(async (manager) => {
      for (const dto of dtos) {
        const newQuestion = new QuestionModel();
        newQuestion.text = dto.text;
        newQuestion.answerMode = dto.answerMode ?? AnswerModeEnum.SINGLE_CHOICE;
        newQuestion.mediaType = dto.mediaType ?? MediaTypeEnum.NONE;

        newQuestion.answers = dto.answers.map((answer) => {
          const newAnswer = new AnswerModel();
          newAnswer.text = answer.text;
          newAnswer.isCorrectAnswer = answer.isCorrectAnswer ?? false;
          newAnswer.question = newQuestion;
          return newAnswer;
        });

        await manager.save(newQuestion);
      }
    });
  }

  async updateQuestion(id: UUID, dto: UpdateQuestionDto): Promise<void> {
    await this.dataSource.transaction(async (manager) => {
      const question = await manager.findOneBy(QuestionModel, { id });
      if (!question) {
        throw new NotFoundException(`Question with id ${id} not found`);
      }

      if (dto.text !== null) question.text = dto.text;
      if (dto.answerMode !== null) question.answerMode = dto.answerMode;
      if (dto.mediaType !== null) question.mediaType = dto.mediaType;

      if (dto.answers !== null) {
        question.answers = dto.answers.map((answer) => {
          const newAnswer = new AnswerModel();
          if (answer.text !== null) newAnswer.text = answer.text;
          newAnswer.isCorrectAnswer = answer.isCorrectAnswer ?? false;
          newAnswer.question = question;
          return newAnswer;
        });
      }

      await manager.save(question);
    });
  }

  async deleteQuestion(id: UUID): Promise<void> {
    await this.questionRepository.delete({ id });
  }
}
